faster_raster.go

```diff
@@ -366,57 +383,85 @@ func (r *Rasterizer) calculateScaleForDocument(pageCount int) {
 	}
 }
 
+func sendErrorReply(req *RasterRequest, filename string, err error) {
+	select {
+	case req.ReplyChan <- &RasterReply{err: err}:
+		// nothing
+	default:
+		log.Warnf("Failed to send reply for %q page %d", filename, req.PageNumber)
+	}
+}
+
+// runCancellableOperation sets cookie.abort to 1 if the caller bailed out early to
+// instruct MuPDF to terminate the pending operation as soon as possible
+func runCancellableOperation(req *RasterRequest, filename string, fn func(*C.fz_cookie)) error {
+	if req.ctx == nil {
+		// In this case, the caller doesn't want to bail out early
+		fn(nil)
+
+		return nil
+	}
+
+	var cookie C.fz_cookie
+	done := make(chan struct{})
+	go func() {
+		select {
+		case <-req.ctx.Done():
+			cookie.abort = 1
```

Comment:
Is any pthreads locking required around this state modification?

Reply:
I was wondering the same, but it seems that's not needed. See [here](https://github.com/ArtifexSoftware/mupdf/blob/9b183c560ef9183f7d4ab32af56d8bad0b494f36/include/mupdf/fitz/device.h#L219) and [here](https://github.com/ArtifexSoftware/mupdf/blob/9b183c560ef9183f7d4ab32af56d8bad0b494f36/source/tools/muraster.c#L680) for the way they use it. Thinking about it, it's just a value in memory that they read periodically and we have no control over when this read occurs. If we change the value, they'll get the updated version eventually and bail out early if the processing hasn't finished yet.